junebug/api.py

```diff
@@ -338,5 +356,70 @@ def create_router(self, request, body):
         ))
 
     @app.route('/health', methods=['GET'])
+    @inlineCallbacks
     def health_status(self, request):
-        return response(request, 'health ok', {})
+        if self.config.rabbitmq_management_interface:
+
+            ids = yield Channel.get_all(self.redis)
+
+            def get_queues(channel_ids):
+                gets = []
+                for channel_id in channel_ids:
+                    for sub in ['inbound', 'outbound', 'event']:
+
+                        queue_name = "%s.%s" % (channel_id, sub)
+                        url = 'http://%s/api/queues/%s/%s' % (
+                            self.config.rabbitmq_management_interface,
+                            self.amqp_config['vhost'].replace('/', '%2F'),
+                            queue_name
+                        )
+                        http_client = HTTPClient(
+                            self.agent_factory(
+                                self.clock,
+                                pool=self.pool_factory(self.clock)))
+
+                        get = http_client.get(
+                            url, auth=(self.amqp_config['username'],
+                                       self.amqp_config['password']))
+                        get.addCallback(treq.json_content)
+                        gets.append(get)
+
+                return gets
+
+            def return_queue_results(results):
+                queues = []
+                stuck = False
+
+                for result in results:
+                    queue = result[1]
+
+                    if ('messages' in queue):
+                        details = {
+                            'name': queue['name'],
+                            'stuck': False,
+                            'messages': queue.get('messages'),
+                            'rate': queue['messages_details']['rate']
+                        }
+                        if (details['messages'] > 0 and details['rate'] == 0):
+                            stuck = True
+                            details['stuck'] = True
+
+                        queues.append(details)
+
+                status = 'channels ok'
+                code = http.OK
+                if stuck:
+                    status = "channels stuck"
+                    code = http.INTERNAL_SERVER_ERROR
+
+                return response(request, status, queues, code=code)
+
+            outputs = get_queues(ids)
+
+            final = defer.DeferredList(outputs)
```

Comment:
The way we're doing it here, we're making all the requests in parallel. The issue with this is that firstly, the connection pool will only be used for multiple requests to `/health`, so I don't think that it's doing much. 

Secondly, if the amount of requests is greater than the pool size, then those extra connections won't be cached.

Thirdly, I don't think that the management interface will appreciate it much if we're making hundreds of parallel requests to it, at some amount of parallel requests I'm sure that the requests will start failing.

I think we need to limit the amount of requests that we're making at any time, I think we can use twisted's DeferredSemaphore to do this, some nice explanations and examples here: https://patrickmn.com/software/using-twisteds-deferred-semaphore/